Training24/Test.c

```diff
@@ -1,14 +1,163 @@
+   printf ("| Input Array:            ");
+   DisplayArray ((int*)inputs, size);
+   printf ("| Output Array:           ");
+   DisplayArray ((int*)sortedArray, size);
+   if (IsSorted ((int*)sortedArray, size)) printf ("Insertion sort: " GREEN "PASS" RESET "\n");
+   else printf ("Insertion sort: " RED "FAIL" RESET "\n");
+   // Find the element in the sorted array
+   int index = BinarySearch ((int*)sortedArray, size, searchElement);
+   if (index != -1) {
+      int originalIndex = sortedArray[index].originalIndex;
+      printf ("Element %d found at sorted index %d (original index %d).\n", searchElement, index, originalIndex);
+   }
+   else printf (RED"Element %d not found.\n"RESET, searchElement);
+   printf ("--------------------------------------------------\n\n");
+}
+
+/// <summary>Tests sorting and searching algorithms with predefined cases.</summary>
+static void TestSort () {
+   struct {
+      int input[50], searchElement, size;
+   } tests[] = {
+       {{34, 7, 23, 32, 10}, 2, 5},          // Search for an element not in the array
+       {{5, 3, 8, 1, 2, 6}, 8, 6},
+       {{11, 2}, 11, 2},
+       {{12, 11, 13, 5}, 5, 4},
+       {{-1, -2, -3, -4, -5, -6, -7}, -3, 7},
+       {{99, 3, 14}, 14, 3},
+       {{-10, 0, 5, 2}, 0, 4},
+       {{1, 2, 3, 4, 5}, 6, 5}               // Search for an element not in the array
+   };
+   int numTests = sizeof (tests) / sizeof (tests[0]);
+   for (int i = 0; i < numTests; i++) {
+      // Declare an array of IndexedElement to hold the indexed inputs for the current test case.
+      IndexedElement indexedInput[50];
+      for (int j = 0; j < tests[i].size; j++) {
+         indexedInput[j].value = tests[i].input[j];
+         indexedInput[j].originalIndex = j;
+      }
+      PrintTestCase (i + 1, indexedInput, tests[i].size, tests[i].searchElement);
+   }
+}
+
+/// <summary>User input for array creation, sorting, and element searching.</summary>
+static void ManualArrayInput () {
+   char continueInput = 'y';
+   while (tolower (continueInput) == 'y') {
+      int arr[50], num;
+      char buffer[100];
+      printf ("\nEnter the number of elements (maximum length of the array is 50): ");
+      fgets (buffer, sizeof (buffer), stdin);
+      num = atoi (buffer);
+      if (num > 50 || num < 1) {
+         printf ("Invalid number of elements. Please enter a number between 1 and 50.\n");
+         continue;
+      }
+      printf ("Enter %d integer numbers:\n", num);
+      for (int i = 0; i < num; i++) {
+         fgets (buffer, sizeof (buffer), stdin);
+         arr[i] = atoi (buffer);
+         if (!IsValidInteger (buffer)) {
+            printf ("Invalid input. Please enter a valid integer.\n");
+            i--; // Decrement i to repeat this iteration for the same index
+            continue;
+         }
+         arr[i] = atoi (buffer);
+      }
+      // Create an array of IndexedElements for sorting
+      IndexedElement indexedArray[50];

```

Comment:
Similar to initial comment, use array itself not necessary to store & display initial index.